src/Cms/Environment.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace Kirby\Cms;
+
+use Kirby\Exception\InvalidArgumentException;
+use Kirby\Filesystem\F;
+use Kirby\Http\Server;
+use Kirby\Http\Uri;
+
+/**
+ * The environment object takes care of
+ * secure host and base URL detection, as
+ * well as loading the dedicated
+ * environment options.
+ *
+ * @package   Kirby Cms
+ * @author    Bastian Allgeier <bastian@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+class Environment
+{
+    /**
+     * @var string
+     */
+    protected $root;
+
+    /**
+     * @var \Kirby\Http\Uri
+     */
+    protected $uri;
+
+    /**
+     * @param string $root
+     * @param bool|string|array|null $allowed
+     */
+    public function __construct(string $root, $allowed = null)
+    {
+        $this->root = $root;
+
+        // the current URL should be detected via possibly insecure HOST headers
+        if ($allowed === true || $allowed === null) {
+            Server::hosts(['*']);
+            $this->uri = Uri::index();
+
+        // the current URL should be detected via the server name
+        } elseif ($allowed === false) {
+            Server::hosts([]);
+            $this->uri = Uri::index();
+
+        // the current URL is predefined and not detected automatically
+        } elseif (is_string($allowed) === true) {
+            $this->uri = new Uri($allowed);
+            Server::hosts([$this->uri->host()]);
+
+        // the current URL should be auto detected from a host allowlist
+        } elseif (is_array($allowed) === true) {
+            foreach ($allowed as $url) {
+                $host = (new Uri($url))->host();
```

Comment:
Will this preserve `fnmatch` patterns?

Reply:
To be honest, the patterns don't really work at all in our url config option. How would they be translated back to proper URLs? I think it's nice to have them in the server class, but in this case they don't have any effect.